src/java.base/share/classes/jdk/internal/foreign/abi/CapturableState.java

```diff
@@ -30,66 +30,79 @@
 import java.lang.foreign.MemoryLayout;
 import java.lang.foreign.StructLayout;
 import java.lang.foreign.ValueLayout;
+import java.util.ArrayList;
 import java.util.List;
-import java.util.stream.Collectors;
-import java.util.stream.Stream;
+import java.util.Map;
 
 import static java.lang.foreign.ValueLayout.JAVA_INT;
 
-public enum CapturableState {
-    GET_LAST_ERROR    ("GetLastError",    JAVA_INT, 1 << 0, OperatingSystem.isWindows()),
-    WSA_GET_LAST_ERROR("WSAGetLastError", JAVA_INT, 1 << 1, OperatingSystem.isWindows()),
-    ERRNO             ("errno",           JAVA_INT, 1 << 2, true);
+public final class CapturableState {
 
-    public static final StructLayout LAYOUT = MemoryLayout.structLayout(
-        supportedStates().map(CapturableState::layout).toArray(MemoryLayout[]::new));
-    public static final List<CapturableState> BY_ORDINAL = List.of(values());
+    public static final StructLayout LAYOUT;
+    private static final Map<String, CapturableState> LOOKUP;
 
     static {
-        assert (BY_ORDINAL.size() < Integer.SIZE); // Update LinkerOptions.CaptureCallState
+        List<CapturableState> supported;
+
+        if (OperatingSystem.isWindows()) {
+            supported = List.of(
+                    new CapturableState("GetLastError",    JAVA_INT, 1 << 0),
+                    new CapturableState("WSAGetLastError", JAVA_INT, 1 << 1),
+                    new CapturableState("errno",           JAVA_INT, 1 << 2)
+            );
+        } else {
+            supported = List.of(new CapturableState("errno", JAVA_INT, 1 << 2));
+        }
+
+        MemoryLayout[] stateLayouts = new MemoryLayout[supported.size()];
+        @SuppressWarnings({"unchecked", "rawtypes"})
+        Map.Entry<String, CapturableState>[] entries = new Map.Entry[supported.size()];
```

Comment:
```suggestion
        CapturableState[] supported;

        if (OperatingSystem.isWindows()) {
            supported = new CapturableState[] {
                    new CapturableState("GetLastError", JAVA_INT, 1 << 0),
                    new CapturableState("WSAGetLastError", JAVA_INT, 1 << 1),
                    new CapturableState("errno", JAVA_INT, 1 << 2)
            };
        } else {
            supported = new CapturableState[] {new CapturableState("errno", JAVA_INT, 1 << 2)};
        }

        MemoryLayout[] stateLayouts = new MemoryLayout[supported.length];
        @SuppressWarnings({"unchecked", "rawtypes"})
        Map.Entry<String, CapturableState>[] entries = new Map.Entry[supported.length];
```
Here it might be better to replace List with an array

Reply:
I think I might ask @JornVernee for opinion as that is the original author.

Reply:
Either one works I think. Array might be a bit more efficient. Don't think it matters much though